refactor!: expanding return value of getLogsByTags

benesjan opened the pull request:
Fixes F-231
Fixes #9789

In this PR I modify the the return value of our "get logs by tags" endpoints such that we don't need to perform another request for tx hash in PXE's log service. Since I was already modifying it I also dropped unused values from the type and made it a bit less messy.

I added here migration notes for all the changes I did to the endpoint not only in this PR but also in PRs down the stack.

benesjan commented:
* **#19290** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19290?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19289** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19289?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19233** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19233?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19233?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#19125** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19125?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19161** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19161?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19158** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19158?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19157** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19157?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19030** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19030?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17611** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17611?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17568** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17568?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17534** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17534?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17445** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17445?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17451** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17451?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on docs/docs-developers/docs/resources/migration_notes.md:
```diff
@@ -9,26 +9,34 @@ Aztec is in full-speed development. Literally every version breaks compatibility
 ### [AVM] Gas cost multipliers for public execution to reach simulation/proving parity
 
 Gas costs for several AVM opcodes have been adjusted with multipliers to better align public simulation costs with actual proving costs.
 
```
Just run formatting below.

benesjan commented on yarn-project/archiver/src/archiver/kv_archiver_store/log_store.ts:
```diff
@@ -57,21 +57,16 @@ export class LogStore {
    * @param block - The L2 block to extract logs from.
    * @returns An object containing the private and public tagged logs for the block.
    */
-  async #extractTaggedLogsFromBlock(block: L2BlockNew) {
```
`blockHash` was dropped from `TxScopedL2Log` which results in this no longer needing to be async.

mverzilli replied:
Why are we dropping it though? The point of adding it was that if we just include `blockNumber` but not `blockHash`, there's no cheap way for someone consuming the node API to know if the logs are obsolete.

Internally, the archiver is guaranteed to have only currently not-reorg'd blocks, so it's not a problem, but from the outside there's no way to keep track of that. So I think dropping this is a step back in that sense (note that PXE might not need this, but there are other users, especially for the public version). Related: https://github.com/AztecProtocol/aztec-packages/issues/16245

mverzilli replied:
Discussed in daily and we concluded that `blockHash` is not needed here, because:

1. The main purpose of the endpoints that use `TxScopedL2Log`  is to serve PXE.
2. Other clients of Aztec Node should instead use `getPublicLogs`, where the `blockHash` is included in the response.

benesjan replied:
As discussed in scrum the block hash value is fine to drop here because externals are not expected to use this API and in our use cases it's not needed as we have other methods of dealing with reorgs.

benesjan commented on yarn-project/pxe/src/logs/log_service.ts:
```diff
@@ -88,19 +88,11 @@ export class LogService {
-      txEffect.data.noteHashes,
-      txEffect.data.nullifiers[0],
+      scopedLog.noteHashes,
+      scopedLog.firstNullifier,
```
Cute, no? 🥰

benesjan commented on yarn-project/pxe/src/logs/log_service.test.ts:
```diff
@@ -90,106 +81,17 @@ describe('LogService', () => {
     });
   });
-
-  describe('getPublicLogByTag', () => {
```
I made the `getPublicLogByTag` function private in the log service and it got greatly simplified by this PR that I don't think we need these tests anymore.

benesjan commented on yarn-project/pxe/src/logs/log_service.ts:
```diff
@@ -72,8 +71,7 @@ export class LogService {
 
-  // TODO(F-231): delete this function and implement this behavior in the node instead
-  public async getPublicLogByTag(tag: Tag, contractAddress: AztecAddress): Promise<PublicLogWithTxData | null> {
+  async #getPublicLogByTag(tag: Tag, contractAddress: AztecAddress): Promise<PublicLogWithTxData | null> {
```
This function `getPrivateLogByTag` were not used out of the `LogService` so I made them private.

I think eventually we might want to drop these functions but kept them around for now due to the linked issue #11627 below.

benesjan commented on yarn-project/stdlib/src/tx/tx_effect.ts:
```diff
@@ -214,12 +214,14 @@ export class TxEffect {
   } = {}): Promise<TxEffect> {
     const count = (max: number, num?: number) => num ?? Math.min(maxEffects ?? randomInt(max), max);
+    // Every tx effect must have at least 1 nullifier (the first nullifier is used for log indexing)
+    const countNullifiers = (max: number, num?: number) => Math.max(1, count(max, num));
```
Needed to update this to make archiver tests pass. It's correct to have min be 1 here because all the tx always have at least 1 nullifier.

mverzilli commented on yarn-project/stdlib/src/logs/tx_scoped_l2_log.ts:
```diff
@@ -1,114 +1,96 @@
+    public logData: Fr[],
+    /*
+     * The note hashes from the tx effect
+     */
+    public noteHashes: Fr[],
+    /*
+     * The first nullifier from the tx effect
+     */
+    public firstNullifier: Fr,
```
Would you share some context on what's the motivation to add `firstNullifier`? Without context it feels ad-hoc

mverzilli replied:
I saw a few comments in log_store... it would be helpful to add one or two sentences here explaining why it is useful for it to be here

benesjan replied:
Added there this:
```
    /*
     * The first nullifier from the tx effect. Used for nonce discovery when processing notes from logs.
     *
     * (Note nonces are computed as `hash(firstNullifier, noteIndexInTx)`.)
     */
```

mverzilli reviewed (changes requested):
Great job!

I just would like to sync on the decision to drop the block hash from the `TxScopedL2Log` since I think there's good reasons for it to be there

benesjan commented on yarn-project/end-to-end/src/composed/e2e_token_bridge_tutorial_test.test.ts:
```diff
@@ -229,5 +229,5 @@ describe('e2e_cross_chain_messaging token_bridge_tutorial_test', () => {
     // docs:end:l1-withdraw
     expect(newL1Balance).toBe(withdrawAmount);
-  }, 90000);
+  }, 110_000);
```
This timed out in merge queue. I think this PR did not affect but the one down the stack did. As discussed on [slack](https://aztecprotocol.slack.com/archives/C07R3GTKJ49/p1767158343532519?thread_ts=1766541757.471949&cid=C07R3GTKJ49) the new log sync algo is a bit slower because we have WINDOW_LEN of 20 and not 10 like before.

benesjan committed:
refactor!: expanding return value of getLogsByTags

Fixes F-231
Fixes #9789

In this PR I modify the the return value of our "get logs by tags" endpoints such that we don't need to perform another request for tx hash in PXE's log service. Since I was already modifying it I also dropped unused values from the type and made it a bit less messy.

I added here migration notes for all the changes I did to the endpoint not only in this PR but also in PRs down the stack.
